src/dataflow/src/sink/kafka.rs

```diff
@@ -85,39 +89,65 @@ pub fn kafka<G>(
     );
+    let mut queue: VecDeque<(Row, Diff)> = VecDeque::new();
+    let mut vector = Vec::new();
+    let mut encoded_buffer = None;
 
     let name = format!("kafka-{}", id);
     stream.sink(
         Exchange::new(move |_| sink_hash),
         &name.clone(),
         move |input| {
-            // TODO(rkhaitan): can we use a better execution model where we
-            // limit the number of inputs we consume at a time?
             input.for_each(|_, rows| {
-                for (row, _time, diff) in rows.iter() {
-                    let diff_pair = if *diff < 0 {
+                rows.swap(&mut vector);
+
+                for (row, _time, diff) in vector.drain(..) {
+                    queue.push_back((row, diff));
+                }
+            });
+
+            for _ in 0..KAFKA_SINK_FUEL {
+                let (encoded, count) = if let Some((encoded, count)) = encoded_buffer.take() {
+                    (encoded, count)
+                } else if queue.is_empty() {
+                    return;
+                } else {
+                    let (row, diff) = queue.pop_front().expect("queue known to be nonempty");
+
+                    if diff == 0 {
+                        // Explicitly refuse to send no-op records
+                        continue;
+                    };
+
+                    let diff_pair = if diff < 0 {
                         DiffPair {
-                            before: Some(row),
+                            before: Some(&row),
                             after: None,
                         }
                     } else {
                         DiffPair {
                             before: None,
-                            after: Some(row),
+                            after: Some(&row),
                         }
                     };
+
                     let buf = encoder.encode_unchecked(connector.schema_id, diff_pair);
-                    for _ in 0..diff.abs() {
-                        let record = BaseRecord::<&Vec<u8>, _>::to(&connector.topic).payload(&buf);
-                        if let Err((e, _)) = producer.inner().send(record) {
-                            // TODO(rkhaitan): can we yield and retry sending
-                            // if the error is specifically a Kafka backpressure
-                            // error?
-                            error!("unable to produce in {}: {}", name, e);
-                        }
+                    (buf, diff.abs())
+                };
+
+                let record = BaseRecord::<&Vec<u8>, _>::to(&connector.topic).payload(&encoded);
+                if let Err((e, _)) = producer.inner().send(record) {
+                    error!("unable to produce in {}: {}", name, e);
+                    if let KafkaError::MessageProduction(RDKafkaError::QueueFull) = e {
+                        encoded_buffer = Some((encoded, count));
+                        return;
```

Comment:
I'm not familiar with the class of Kafka errors, but it looks like if we reach one other than `QueueFull` we will spam the heck out of everyone. Maybe best to return in that case too?

Reply:
I'm not sure either which set of errors is worth retrying for vs which is not. this behavior preserves current functionality with the exception of QueueFull

Reply:
I'm a bit confused about the error case. It seems like for `QueueFull` we return without decrementing the count (good), whereas for any other error we mark the record as sent and continue. There appear to be other sorts of errors like `OperationTimedOut` that we should retry, and others that we should probably do something a bit more assertive. I don't know that this PR is the right place to enumerate all of those, but we should probably make a note somewhere that this error means data loss occurred.

Reply:
Another option would just be to retry everything after 60s, or whatever, which has the property that transient errors can recover and persistent failures just turn in to memory leaks. I'm not sure which case worries us more, but my default guess would be avoid the data loss to start.